File: L129-sum_root_to_leaf_numbers.py
# https://leetcode.com/problems/sum-root-to-leaf-numbers/
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        def makeSum(root, cur, wholeSum):
            if (not root):
                return wholeSum
            cur = cur * 10 + root.val
            if (not root.left and not root.right):
                wholeSum += cur
                cur = 0
                return wholeSum
            wholeSum = makeSum(root.left, cur, wholeSum)
            wholeSum = makeSum(root.right, cur, wholeSum)
            return wholeSum
        return makeSum(root, 0, 0)

File: L129-SumRootToLeafNumbers.java
// https://leetcode.com/problems/sum-root-to-leaf-numbers/
class Solution {
    public int makeSum(TreeNode root,int cur,int sum) {
        if (root == null) {
            return sum;
        }
        cur = cur * 10 + root.val;
        if (root.left == null && root.right == null) {
            sum += cur;
            cur = 0;
        }
        sum = makeSum(root.left, cur, sum);
        sum = makeSum(root.right, cur, sum);
        return sum;
    }
    public int sumNumbers(TreeNode root) {
        return makeSum(root, 0, 0);
    }
}
